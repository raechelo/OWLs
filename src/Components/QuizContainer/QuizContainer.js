import React, { Component } from 'react';
import Question from '../Question/Question';
import Answer from '../Answer/Answer';
import Player from '../Player/Player';
import Filter from '../Filter/filter';
import Popup from '../Popup/Popup';

class QuizContainer extends Component {
  constructor() {
    super();
    this.state = {
      category: '',
      questions: [],
      currentQuestion: {},
      incorrect: [],
      house: '',
      h: 0,
      r: 0,
      g: 0,
      s: 0
      }
    }
  

  randomizeQuestions = () => {
  let shuffledQs = this.props.studySet.sort(() => 0.5 - Math.random())
  let currentQ = shuffledQs.pop()

    this.setState( {
      questions: shuffledQs,
      currentQuestion: currentQ,
      incorrect: this.props.incorrect,
      house: this.props.house
    } )
  }

  incorrectAns = () => {
    if (!this.state.incorrect.includes(this.state.currentQuestion)) {
      let newState = this.state.incorrect.push(this.state.currentQuestion)
      this.setState( { incorrect: newState } )
      this.props.saveWrongQs();
    }
  }

  addPoints = () => {
    switch (this.state.house) {
      case 'gryffindor' :
        this.setState( { g: this.state.g += 100 } )
        console.log(this.state.g)
      break;
      case 'slytherin':
      this.setState( { s: this.state.s += 100 } )
      break
      case 'hufflepuff':
      this.setState( { h: this.state.h += 100 } )
      break;
      default:
      this.setState( { r: this.state.r += 100 } )
    }
  }

  addOtherPoints = () => {
    switch (this.state.house) {
      case 'gryffindor' :
        this.setState( { 
          r: this.state.r += 100,
          s: this.state.s += 100,
          s: this.state.h += 100
        } )
      break;
      case 'slytherin':
        this.setState( { 
          r: this.state.r += 100,
          s: this.state.g += 100,
          s: this.state.h += 100
        } )
      break
      case 'hufflepuff':
        this.setState( { 
          r: this.state.r += 100,
          s: this.state.s += 100,
          s: this.state.g += 100
        } )
      break;
      case 'ravenclaw':
      this.setState( { 
        r: this.state.h += 100,
        s: this.state.s += 100,
        s: this.state.g += 100
      } )
      default:
      this.setState( { 
        r: this.state.h,
        s: this.state.s,
        s: this.state.g
      } )
    }
  }



  render() {
    const gameBoard =  
    <div className="game-board">
      <Question 
        question={this.state.currentQuestion.question} />
      <Answer
        addOtherPoints={this.addOtherPoints}
        addPoints={this.addPoints}
        saveWrongQs={this.props.saveWrongQs}
        incorrectAns={this.incorrectAns}
        nextQuestion={this.randomizeQuestions} 
        correctAnswer={this.state.currentQuestion.correctAnswer} 
        allAnswers={this.state.currentQuestion.allAnswers} />
    </div>


    return (
      <div className="game-container">
        {this.state.questions.length > 0 
        ? gameBoard 
        : <Popup 
          clearStorage={this.props.clearStorage}
          randomizeQuestions={this.randomizeQuestions}
          studyWrongQs={this.props.studyWrongQs}
          filterBy={this.filterBy}
         /> }
        <Player 
          house={this.props.house} 
          h={this.state.h} 
          s={this.state.s} 
          g={this.state.g} 
          r={this.state.r} />
      </div>
    )
  }
}

export default QuizContainer;