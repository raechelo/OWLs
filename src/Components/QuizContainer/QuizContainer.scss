.info-popup {
  background-color: $boardBg;
  color: white;
  font-family: $bodyfont;
  font-size: 1.5rem;
  margin: 1rem auto;
  padding: 10px;
  text-align: center;
  width: 65%;

  button {
    font-family: $bodyfont;
    font-size: 1rem;
    padding: 5px 10px;
    cursor: pointer;
  }

  h3 {
    margin: .5rem;
    letter-spacing: 2px;
    text-decoration: underline;
    text-transform: uppercase;
  }
}

.game-board {
  background-color: $boardBg;
  margin: 1rem auto;
  width: $boardW;
}