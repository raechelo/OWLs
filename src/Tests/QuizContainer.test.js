import React from 'react';
import QuizContainer from '../Components/QuizContainer/QuizContainer';
import { shallow } from 'enzyme';

describe('QuizContainer', () => {
  
  let wrapper;

  const mockIncorrect = [];

  const mockSaveWrongQs = jest.fn();

  const mockHouse = 'ravenclaw'

  const data = [
    {
      "question": "In which lifecycle method should you fetch APIs?",
      "correctAnswer": "componentDidMount",
      "allAnswers": ["componentDidMount", "componentWillMount", "componentDoesMount", "componentDidUnmount"],
      "category": "React",
      "id": 1
    },
    {
      "question": "What would be the proper way to add new values to an array stored in state?",
      "correctAnswer": "this.setState({myArray: newValues})",
      "allAnswers": ["this.setState({myArray: newValues})", "this.setState(newValues)", "this.state.myArray.push(newValues)", "myArray.push(newValues)"],
      "category": "React",
      "id": 2
    }
  ];

  beforeEach(() => {
    wrapper = shallow (
      <QuizContainer incorrect={mockIncorrect} studySet={data} saveWrongQs={mockSaveWrongQs} house={mockHouse} />
    )
  });

  it('should match the snapshot wtih all the data passed in', () => {
    wrapper.setState( {questions: []} )
    expect(wrapper).toMatchSnapshot();
    
    wrapper.setState( {questions: data} )
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a proper default state', () => {
    expect(wrapper.state()).toEqual( {category: '', questions: [], currentQuestion: {}, incorrect: [], house: '', h: 0, r: 0, g:0, s:0 } )
  });

  it('should randomize questions', () => {
    expect(wrapper.state('questions')).toEqual([]);
    wrapper.instance().randomizeQuestions();
    expect(wrapper.state('currentQuestion')).toHaveProperty('allAnswers', 'id', 'question', 'correctAnswer', 'category');
    expect(wrapper.state('incorrect')).toEqual(mockIncorrect);
    expect(wrapper.state('house')).toEqual(mockHouse)
  });

  it('should add incorrect questions to the proper array', () => {
    expect(wrapper.state('incorrect')).toEqual([]);
    wrapper.instance().randomizeQuestions();
    wrapper.instance().incorrectAns();
    expect(wrapper.state('incorrect')).toEqual(1);
  });

  it('should add points to the players house for correct answers', () => {
    wrapper.setState( { house:'gryffindor' } )
    expect(wrapper.state('g')).toEqual( 0 );
    wrapper.instance().addPoints();
    expect(wrapper.state('g')).toEqual(100);
    
    wrapper.setState( { house: 'slytherin' } );
    expect(wrapper.state('s')).toEqual( 0 );
    wrapper.instance().addPoints();
    expect(wrapper.state('s')).toEqual(100);

    wrapper.setState( { house: 'hufflepuff' } );
    expect(wrapper.state('h')).toEqual( 0 );
    wrapper.instance().addPoints();
    expect(wrapper.state('h')).toEqual(100);

    wrapper.setState( { house: 'ravenclaw' } );
    expect(wrapper.state('r')).toEqual( 0 );
    wrapper.instance().addPoints();
    expect(wrapper.state('r')).toEqual(100);
  });

  it('should add points to other houses when a user answers incorrectly', () => {
    wrapper.setState( {house: 'gryffindor'} );
    expect(wrapper.state('g')).toEqual(0);
    wrapper.instance().addOtherPoints();
    expect(wrapper.state('r')).toEqual( 100 )
    expect(wrapper.state('s')).toEqual( 100 )
    expect(wrapper.state('h')).toEqual( 100 )

    wrapper.setState( {house: 'slytherin', g: 0, h: 0, s: 0, r: 0} );
    expect(wrapper.state('s')).toEqual(0);
    wrapper.instance().addOtherPoints();
    expect(wrapper.state('r')).toEqual( 100 )
    expect(wrapper.state('g')).toEqual( 100 )
    expect(wrapper.state('h')).toEqual( 100 )

    wrapper.setState( {house: 'hufflepuff', g: 0, h: 0, s: 0, r: 0} );
    expect(wrapper.state('h')).toEqual(0);
    wrapper.instance().addOtherPoints();
    expect(wrapper.state('r')).toEqual( 100 )
    expect(wrapper.state('s')).toEqual( 100 )
    expect(wrapper.state('g')).toEqual( 100 )

    wrapper.setState( {house: 'ravenclaw', g: 0, h: 0, s: 0, r: 0} );
    expect(wrapper.state('r')).toEqual(0);
    wrapper.instance().addOtherPoints();
    expect(wrapper.state('g')).toEqual( 100 )
    expect(wrapper.state('s')).toEqual( 100 )
    expect(wrapper.state('h')).toEqual( 100 )
  });
});
