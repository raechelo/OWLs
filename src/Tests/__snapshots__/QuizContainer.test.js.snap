// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizContainer should match the snapshot wtih all the data passed in 1`] = `
<div
  className="game-container"
>
  <Popup
    randomizeQuestions={[Function]}
  />
  <Player
    g={0}
    h={0}
    house="ravenclaw"
    r={0}
    s={0}
  />
</div>
`;

exports[`QuizContainer should match the snapshot wtih all the data passed in 2`] = `
<div
  className="game-container"
>
  <div
    className="game-board"
  >
    <Question />
    <Answer
      addOtherPoints={[Function]}
      addPoints={[Function]}
      incorrectAns={[Function]}
      nextQuestion={[Function]}
      saveWrongQs={[MockFunction]}
    />
  </div>
  <Player
    g={0}
    h={0}
    house="ravenclaw"
    r={0}
    s={0}
  />
</div>
`;
